benchmark_online.py

```diff
@@ -0,0 +1,175 @@
+
+EXAMPLE:
+export CUDA_VISIBLE_DEVICES=1
+python benchmark_online.py \
+--checkpoint best_checkpoints/JasperNetBig_NovoGrad_lr1e-4_wd1e-3_bs512___\
+finetune_after_self_train_on_2020_11_27_clean_valset_epoch144_iter0335000.pt \
+--device cuda -T 6.0 --benchmark-duration 60 --rps 5.0
+
+OUT:
+initializing model...
+batch [1, 48000] | audio 6.00 sec
+Warming up for 100 iterations...
+Warmup done in 2.7 sec
+Starting 60 second benchmark (300 requests, rps 5.0)...
+avg gap between requests: 195.0 ms
+Latency mean: 33.6 ms, median: 33.1 ms, 90-th percentile: 36.9 ms, 95-th percentile: 46.2 ms,
+99-th percentile: 60.7 ms, max: 73.0 ms | service idle time fraction: 81.7%
+
+ONNX EXAMPLE:
+export CUDA_VISIBLE_DEVICES=1
+python benchmark_online.py \
+--onnx best_checkpoints/JasperNetBig_NovoGrad_lr1e-4_wd1e-3_bs512____\
+rerun_finetune_after_self_train_epoch183_iter0290000.pt.12.fp16_masking_1.7.1.onnx \
+--device cuda -T 6.0 --benchmark-duration 60 --rps 5.0
+
+OUT:
+initializing model...
+initial providers: ['CUDAExecutionProvider', 'CPUExecutionProvider']
+changed providers: ['CUDAExecutionProvider', 'CPUExecutionProvider']
+batch [1, 48000] | audio 6.00 sec
+Warming up for 100 iterations...
+Warmup done in 2.5 sec
+Starting 60 second benchmark (300 requests, rps 5.0)...
+avg gap between requests: 199.1 ms
+Latency mean: 22.7 ms, median: 21.7 ms, 90-th percentile: 23.8 ms, 95-th percentile: 32.3 ms,
+99-th percentile: 42.5 ms, max: 61.2 ms | service idle time fraction: 88.9%
+
+Tests show, that current model on 1 GPU can handle RPS=50 (x100 of our production), with peak latency < 500ms.
+"""
+
+import argparse
+import os
+import math
+import time
+import json
+
+import torch
+import torch.cuda.profiler
+import onnxruntime
+import numpy as np
+
+import models
+import text_processing
+
+
+def main(args):
+	use_cuda = 'cuda' in args.device
+	if use_cuda:
+		print(f'CUDA_VISIBLE_DEVICES: {os.environ.get("CUDA_VISIBLE_DEVICES")!r}')
+	print('initializing model...')
+
+	if args.onnx:
+		# todo: pass dict with provider setting when we will migrate to onnxruntime>=1.7
+		onnxruntime_session = onnxruntime.InferenceSession(args.onnx)
+		model = lambda x: onnxruntime_session.run(None, dict(x=x, xlen=[1.0] * len(x)))
+		load_batch = lambda x: x.numpy()
+		args.sample_rate = 8000
+	else:
+		assert args.checkpoint is not None and os.path.isfile(args.checkpoint)
+		checkpoint = torch.load(args.checkpoint, map_location='cpu')
+		args.model, args.lang, args.sample_rate, args.window_size, args.window_stride, args.window, args.num_input_features = (
+			checkpoint['args'].get(key) for key in
+			['model', 'lang', 'sample_rate', 'window_size', 'window_stride', 'window', 'num_input_features']
+		)
+
+		text_config = json.load(open(args.text_config))
+		text_pipelines = []
+		for pipeline_name in args.text_pipelines:
+			text_pipelines.append(text_processing.ProcessingPipeline.make(text_config, pipeline_name))
+
+		frontend = models.LogFilterBankFrontend(
```

Comment:
please use kwargs for all arguments. Othewise, the code may become silently wrong if we reorder arguments

Reply:
Done